File: src/com/kedzie/vbox/api/INetworkAdapter.java
package com.kedzie.vbox.api;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import android.os.Parcel;
import android.os.Parcelable;

import com.kedzie.vbox.api.jaxb.NetworkAdapterPromiscModePolicy;
import com.kedzie.vbox.api.jaxb.NetworkAdapterType;
import com.kedzie.vbox.api.jaxb.NetworkAttachmentType;
import com.kedzie.vbox.soap.Asyncronous;
import com.kedzie.vbox.soap.KSOAP;
import com.kedzie.vbox.soap.VBoxSvc;

public interface INetworkAdapter extends IManagedObjectRef, Parcelable {
	public static final String BUNDLE = "network";
	static final ClassLoader LOADER = INetworkAdapter.class.getClassLoader();
	
	public static final Parcelable.Creator<INetworkAdapter> CREATOR = new Parcelable.Creator<INetworkAdapter>() {
		public INetworkAdapter createFromParcel(Parcel in) {
			VBoxSvc vmgr =  in.readParcelable(LOADER);
			String id = in.readString();
			Map<String, Object> cache = new HashMap<String, Object>();
			in.readMap(cache, LOADER);
			return (INetworkAdapter) vmgr.getProxy(INetworkAdapter.class, id, cache); 
		}
		public INetworkAdapter[] newArray(int size) {  
			return new INetworkAdapter[size]; 
		}
	};
	
	@KSOAP(cacheable=true) public NetworkAdapterType getAdapterType();
	@Asyncronous public void setAdapterType(@KSOAP("adapterType") NetworkAdapterType adapterType);
	
	@KSOAP(cacheable=true) public Integer getSlot();

	@KSOAP(cacheable=true) public Boolean getEnabled();
	@Asyncronous public void setEnabled(@KSOAP("enabled") boolean enabled);

	@KSOAP(cacheable=true) public String getMACAddress();
	@Asyncronous public void setMACAddress(@KSOAP("MACAddress") String macAddress);

	@KSOAP(cacheable=true) public NetworkAttachmentType getAttachmentType();
	@Asyncronous public void setAttachmentType(@KSOAP("attachmentType") NetworkAttachmentType attachmentType);

	@KSOAP(cacheable=true) public String getBridgedInterface();
	@Asyncronous public void setBridgedInterface(@KSOAP("bridgedInterface") String bridgedInterface);

	@KSOAP(cacheable=true) public String getHostOnlyInterface();
	@Asyncronous public void setHostOnlyInterface(@KSOAP("hostOnlyInterface") String hostOnlyInterface);

	@KSOAP(cacheable=true) public String getInternalNetwork();
	@Asyncronous public void setInternalNetwork(@KSOAP("internalNetwork") String internalNetwork);

	@KSOAP(cacheable=true) public String getNATNetwork();
	@Asyncronous public void setNATNetwork(@KSOAP("NATNetwork") String natNetwork);

	@KSOAP(cacheable=true) public String getGenericDriver();
	@Asyncronous public void setGenericDriver(@KSOAP("genericDriver") String genericDriver);
	
	@KSOAP(cacheable=true) public Boolean getCableConnected();
	@Asyncronous public void setCableConnected(@KSOAP("cableConnected") boolean cableConnected);

	@KSOAP(cacheable=true) public Integer getLineSpeed();
	@Asyncronous public void setLineSpeed(@KSOAP(type="unsignedInt", value="lineSpeed") int lineSpeed);
	
	@KSOAP(cacheable=true) public NetworkAdapterPromiscModePolicy getPromiscModePolicy();
	@Asyncronous public void setPromiscModePolicy(@KSOAP("promiscModePolicy") NetworkAdapterPromiscModePolicy promiscModePolicy);

	@KSOAP(cacheable=true) public Boolean getTraceEnabled();
	@Asyncronous public void setTraceEnabled(@KSOAP("traceEnabled") boolean traceEnabled);
	
	@KSOAP(cacheable=true) public String getTraceFile();
	@Asyncronous public void setTraceFile(@KSOAP("traceFile") String traceFile);
	
	@KSOAP(cacheable=true) public Integer getBootPriority();
	@Asyncronous public void setBootPriority(@KSOAP("bootPriority") int bootPriority);
	
//	@KSOAP(cacheable=true) public IBandwidthGroup getBandwidthGroup();
//	@Asyncronous public void setBandwidthGroup(@KSOAP("bandwidthGroup") IBandwidthGroup bandwidthGroup);
//	KSOAP(cacheable=true) public INATEngine getNATEngine();
	
	@KSOAP(cacheable=true) String getProperty(@KSOAP("key") String key);
	@Asyncronous public void setProperty(@KSOAP("key") String key, @KSOAP("value") String value);
	@KSOAP(cacheable=true) public Map<String, List<String>> getProperties(@KSOAP("names") String names);
}

File: vbox-app/src/main/java/com/kedzie/vbox/api/INetworkAdapter.kt
package com.kedzie.vbox.api

import android.os.Parcelable
import com.kedzie.vbox.api.jaxb.NetworkAdapterPromiscModePolicy
import com.kedzie.vbox.api.jaxb.NetworkAdapterType
import com.kedzie.vbox.api.jaxb.NetworkAttachmentType
import com.kedzie.vbox.app.Utils
import com.kedzie.vbox.soap.Cacheable
import com.kedzie.vbox.soap.Ksoap
import com.kedzie.vbox.soap.KsoapProxy
import java.io.IOException
import java.util.*

@KsoapProxy
@Ksoap
interface INetworkAdapter : IManagedObjectRef, Parcelable {
    companion object {
        const val BUNDLE = "networkadapter"
    }

    @Cacheable("adapterType")
    suspend fun getAdapterType(): NetworkAdapterType
    suspend fun setAdapterType(@Cacheable("adapterType") adapterType: NetworkAdapterType)

    @Cacheable("Slot")
	suspend fun getSlot(): Int;

    @Cacheable("enabled")
	suspend fun getEnabled(): Boolean
    suspend fun setEnabled(@Cacheable("enabled") enabled: Boolean)

    @Cacheable("MACAddress")
	suspend fun getMACAddress(): String
    suspend fun setMACAddress(@Cacheable("MACAddress") MACAddress: String)

    @Cacheable("attachmentType")
	suspend fun getAttachmentType(): NetworkAttachmentType
    suspend fun setAttachmentType(@Cacheable("attachmentType") attachmentType: NetworkAttachmentType)

    @Cacheable("bridgedInterface")
	suspend fun getBridgedInterface(): String
    suspend fun setBridgedInterface(@Cacheable("bridgedInterface") bridgedInterface: String)

    @Cacheable("hostOnlyInterface")
	suspend fun getHostOnlyInterface(): String
    suspend fun setHostOnlyInterface(@Cacheable("hostOnlyInterface") hostOnlyInterface: String)

    @Cacheable("internalNetwork")
	suspend fun getInternalNetwork(): String
    suspend fun setInternalNetwork(@Cacheable("internalNetwork") internalNetwork: String)

    @Cacheable("NATNetwork")
	suspend fun getNATNetwork(): String
    suspend fun setNATNetwork(@Cacheable("NATNetwork") natNetwork: String)

    @Cacheable("genericDriver")
	suspend fun getGenericDriver(): String
    suspend fun setGenericDriver(@Cacheable("genericDriver") genericDriver: String)

    @Cacheable("cableConnected")
	suspend fun getCableConnected(): Boolean
    suspend fun setCableConnected(@Cacheable("cableConnected") cableConnected: Boolean)

    @Cacheable("lineSpeed")
	suspend fun getLineSpeed(): Int;
    suspend fun setLineSpeed(@Cacheable("lineSpeed") @Ksoap(type = "unsignedInt") lineSpeed: Int)

    @Cacheable("promiscModePolicy")
	suspend fun getPromiscModePolicy(): NetworkAdapterPromiscModePolicy;
    suspend fun setPromiscModePolicy(@Cacheable("promiscModePolicy") promiscModePolicy: NetworkAdapterPromiscModePolicy)

    @Cacheable("traceEnabled")
	suspend fun getTraceEnabled(): Boolean;
    suspend fun setTraceEnabled(@Cacheable("traceEnabled") traceEnabled: Boolean)

    @Cacheable("traceFile")
	suspend fun getTraceFile(): String;
    suspend fun setTraceFile(@Cacheable("traceFile") traceFile: String)

    @Cacheable("bootPriority")
	suspend fun getBootPriority(): Int;
    suspend fun setBootPriority(@Cacheable("bootPriority") bootPriority: Int)

    //	@Cacheable("BandwidthGroup") suspend fun getBandwidthGroup(): IBandwidthGroup;
//	@Asyncronous public void setBandwidthGroup(@Ksoap("bandwidthGroup") IBandwidthGroup bandwidthGroup);
//	Ksoap(cacheable=true) public INATEngine getNATEngine();

    suspend fun getProperty(key: String): String

    suspend fun setProperty(key: String, value: String)

    suspend fun getProperties(names: String): Map<String, List<String>>
}

/**
 * Load network adapter properties
 * @param names  Property names to load, or empty for all
 * @return    properties
 */
suspend fun INetworkAdapter.getProperties(vararg names: String): Properties {
    val nameString = StringBuffer()
    for (name in names)
        Utils.appendWithComma(nameString, name)

    val map = getProperties(nameString.toString())
    val returnNames = map.get("returnNames")
    val values = map.get("returnval")
    val properties = Properties()
    for (i in returnNames!!.indices)
        properties[returnNames!![i]] = values!![i]
    return properties
}